bot/exts/moderation/clean.py

```diff
@@ -110,23 +107,46 @@ def _validate_input(
         if users and bots_only:
             raise BadArgument("Marked as bots only, but users were specified.")
 
-        # This is an implementation error rather than user error.
-        if second_limit and not first_limit:
-            raise ValueError("Second limit specified without the first.")
-
     @staticmethod
     async def _send_expiring_message(ctx: Context, content: str) -> None:
         """Send `content` to the context channel. Automatically delete if it's not a mod channel."""
         delete_after = None if is_mod_channel(ctx.channel) else MESSAGE_DELETE_DELAY
         await ctx.send(content, delete_after=delete_after)
 
+    @staticmethod
+    def _channels_set(
+            channels: CleanChannels, ctx: Context, first_limit: CleanLimit, second_limit: CleanLimit
+    ) -> set[TextChannel]:
+        """Standardize the input `channels` argument to a usable set of text channels."""
+        # Default to using the invoking context's channel or the channel of the message limit(s).
+        if not channels:
+            # Input was validated - if first_limit is a message, second_limit won't point at a different channel.
+            if isinstance(first_limit, Message):
+                channels = {first_limit.channel}
+            elif isinstance(second_limit, Message):
+                channels = {second_limit.channel}
+            else:
+                channels = {ctx.channel}
+        else:
+            if channels == "*":
+                channels = {
+                    channel for channel in ctx.guild.channels
+                    if isinstance(channel, TextChannel)
+                    # Assume that non-public channels are not needed to optimize for speed.
```

Comment:
I understand why it's done, but what happens if we have "public" channels that are still locked behind a role such as during codejams.

Reply:
The participants channel stays public, just not writeable. If we get to the point of cleaning a user from the team leaders channel, I think a more focused action will be made anyway, likely including the disqualification of the participant. Specifying such a channel explicitly in that case seems like an acceptable edge case. And if it stops being acceptable we can consider adding another shortcut for actually-all channels.